File: examples/vrp/src/persistence/domain_builder.rs


use crate::domain::{VehicleRoutingPlan, Customer, Vehicle};
use std::collections::HashMap;
use std::fs::File;
use std::io::{prelude::*, BufReader, Lines};
use regex::Regex;
use greyjack::domain::DomainBuilderTrait;
use greyjack::utils::math_utils::round;
use greyjack::score_calculation::scores::HardMediumSoftScore;
use serde_json::*;
use polars::datatypes::AnyValue;


#[derive(Clone)]
pub struct DomainBuilder {
    vrp_file_path: String,
}

impl DomainBuilderTrait<VehicleRoutingPlan> for DomainBuilder {
    fn build_domain_from_scratch(&self) -> VehicleRoutingPlan {
        let (metadata, mut customers_vec, readed_distance_matrix, demand_info, depot_info) = Self::read_vrp_file(&self.vrp_file_path);

        assert_eq!(customers_vec.len(), demand_info.len(), "Customers or demands have been readed incorrect");

        let mut time_window_task = false;
        for i in 0..customers_vec.len() {

            assert_eq!(customers_vec[i].id, demand_info[i][0] as usize, "Invalid customer to demand mapping");
            //customers_vec[i].id = demand_info[i][0] as usize;
            customers_vec[i].demand = demand_info[i][1];

            if demand_info[i].len() == 5 {
                time_window_task = true;
                customers_vec[i].time_window_start = demand_info[i][2];
                customers_vec[i].time_window_end = demand_info[i][3];
                customers_vec[i].service_time = demand_info[i][4];
            }
        }
        
        let mut distance_matrix= Vec::new();
        match readed_distance_matrix {
            None => distance_matrix = Self::build_distance_matrix(&customers_vec),
            Some(dm) => {
                for i in 0..customers_vec.len() {
                    let mut distances_to_other_customers: HashMap<String, f64> = HashMap::new();
                    for j in 0..customers_vec.len() {
                        let current_distance = round(dm[i][j], 3);
                        let to_customer_name = customers_vec[j].name.clone();
                        distances_to_other_customers.insert(to_customer_name, current_distance);
                    }
                    customers_vec[i].distances_to_other_customers_map = Some(distances_to_other_customers);
                    distance_matrix = dm.clone();
                }
            }
        }
        for i in 0..customers_vec.len() {
            for j in 0..customers_vec.len() {
                distance_matrix[i][j] = round(distance_matrix[i][j], 3);
            }
        }

        let k_vehicles = metadata["vehicles_count"].parse().unwrap();
        let n_depots = depot_info.len();
        let max_stops = customers_vec.len() - n_depots;
        let capacity = metadata["vehicles_capacity"].parse().unwrap();

        let mut vehicles: Vec<Vehicle> = Vec::new();
        for i in 0..k_vehicles {
            let depot_vec_id = i % n_depots;
            let depot = customers_vec[depot_vec_id].clone();
            let work_day_start = depot.time_window_start;
            let work_day_end = depot.time_window_end;
            let customers: Vec<Customer> = Vec::new();
            let vehicle = Vehicle::new(depot, customers, depot_vec_id, work_day_start, work_day_end, capacity, max_stops);
            vehicles.push( vehicle );
        }

        let mut depot_vec: Vec<Customer> = Vec::new();
        for i in 0..n_depots {
            depot_vec.push(customers_vec[i].clone());
        }

        let domain_model = VehicleRoutingPlan::new(
            metadata["dataset_name"].clone(), vehicles, customers_vec, 
            distance_matrix, depot_vec, time_window_task
        );

        return domain_model;
    }

    fn build_from_solution(&self, solution: &Value) -> VehicleRoutingPlan {

        let mut domain = self.build_domain_from_scratch();
        let solution: (Vec<(String, AnyValue)>, HardMediumSoftScore) = from_value(solution.clone()).unwrap();
        let solution = solution.0;

        for i in (0..solution.len()).step_by(2) {
            
            // TODO: think about how to preserve order in cotwin entity representations
            let (encoded_vehicle_id, encoded_customer_id);
            if solution[i].0.contains("vehicle") {
                (encoded_vehicle_id, encoded_customer_id) = (i, i+1);
            } else {
                (encoded_vehicle_id, encoded_customer_id) = (i+1, i);
            }

            let vehicle_id: usize;
            match solution[encoded_vehicle_id].1 {
                AnyValue::Int64(v_i) => vehicle_id = v_i as usize,
                _ => panic!("Dragons")
            }

            let customer_id;
            match solution[encoded_customer_id].1 {
                AnyValue::Int64(c_i) => customer_id = c_i as usize,
                _ => panic!("Dragons")
            }

            let current_customer = domain.customers_vec[customer_id].clone();
            domain.vehicles[vehicle_id].customers.push( current_customer );
        }

        return domain;
    }
}

impl DomainBuilder {

    pub fn new(vrp_file_path: &str) -> Self {
        Self {
            vrp_file_path: vrp_file_path.to_string(),
        }
    }

    pub fn read_vrp_file(vrp_file_path: &str) -> (HashMap<String, String>, Vec<Customer>, Option<Vec<Vec<f64>>>, Vec<Vec<u64>>, Vec<u64>) {

        let file_pointer = File::open(vrp_file_path).expect(format!("Failed to read {}", vrp_file_path).as_str());
        let file_reader = BufReader::new(file_pointer);
        let mut lines_iter = file_reader.lines().into_iter();

        let metadata: HashMap<String, String> = Self::read_metadata(&mut lines_iter);
        let customers_vec = Self::read_customers_vec(&mut lines_iter);
        
        let mut readed_distance_matrix: Option<Vec<Vec<f64>>> = None;
        if metadata["distance_type"].contains("EUC_2D") == false {
            readed_distance_matrix = Self::read_distance_matrix(&mut lines_iter);
        }

        let demand_info: Vec<Vec<u64>> = Self::read_customers_demand_info(&mut lines_iter);
        let depot_info: Vec<u64> = Self::read_depot_info(&mut lines_iter);

        return (metadata, customers_vec, readed_distance_matrix, demand_info, depot_info);

    }

    fn read_metadata(lines_iter: &mut Lines<BufReader<File>>) -> HashMap<String, String> {

        let mut metadata: HashMap<String, String> = HashMap::new();
        loop {
            let readed_line = lines_iter.next().expect("Failed to read next line 1").expect("Failed to read next line 2");
            
            if readed_line.contains("NODE_COORD_SECTION") {
                break;
            }
            
            let readed_line = readed_line.replace("\n", "");
            if readed_line.contains("NAME") {
                let task_name: Vec<String> = readed_line.split(" ").into_iter().map(|x| x.to_string()).collect();
                let task_name = task_name[task_name.len() - 1].clone();
                let vehicles_count: Vec<String> = task_name.split("-").into_iter().map(|x| x.to_string()).collect();
                let vehicles_count = vehicles_count[vehicles_count.len() - 1].clone();
                let vehicles_count = vehicles_count.replace("k", "");
                metadata.insert("vehicles_count".to_string(), vehicles_count);
            }

            if readed_line.contains("TYPE") {
                let task_type: Vec<String> = readed_line.split(" ").into_iter().map(|x| x.to_string()).collect();
                let task_type = task_type[task_type.len() - 1].clone();
                metadata.insert("task_type".to_string(), task_type);
            }

            if readed_line.contains("NAME") {
                let dataset_name: Vec<String> = readed_line.split(" ").into_iter().map(|x| x.to_string()).collect();
                let dataset_name = dataset_name[dataset_name.len() - 1].clone();
                metadata.insert("dataset_name".to_string(), dataset_name);
            }

            if readed_line.contains("EDGE_WEIGHT_TYPE") {
                let distance_type: Vec<String> = readed_line.split(" ").into_iter().map(|x| x.to_string()).collect();
                let distance_type = distance_type[distance_type.len() - 1].clone();
                metadata.insert("distance_type".to_string(), distance_type);
            }

            if readed_line.contains("CAPACITY") {
                let vehicles_capacity: Vec<String> = readed_line.split(" ").into_iter().map(|x| x.to_string()).collect();
                let vehicles_capacity = vehicles_capacity[vehicles_capacity.len() - 1].clone();
                metadata.insert("vehicles_capacity".to_string(), vehicles_capacity);
            }
        }

        return metadata;
    }

    fn read_customers_vec(lines_iter: &mut Lines<BufReader<File>>) -> Vec<Customer> {

        let mut customers_counter = 0;
        let mut customers_vec: Vec<Customer> = Vec::new();
        loop {
            let mut readed_line = lines_iter.next().expect("Failed to read next line 1").expect("Failed to read next line 2");
            
            if readed_line.contains("EOF") || readed_line.contains("DEMAND_SECTION") {
                break;
            }

            readed_line = readed_line.replace("\n", "");
            readed_line = readed_line.trim().to_string();
            readed_line = Regex::new(r"\s+").unwrap().replace_all(readed_line.clone().as_str(), " ").to_string();

            let customer_parts: Vec<&str> = readed_line.split(" ").collect();
            let id: usize = customer_parts[0].parse().unwrap();
            let latitude: f64 = customer_parts[1].parse().unwrap();
            let longitude: f64 = customer_parts[2].parse().unwrap();
            let name: String;
            if customer_parts.len() > 3 {
                name = customer_parts[3].replace("\n", "");
            } else {
                name = id.to_string();
            }

            let readed_customer = Customer{
                id: id,
                vec_id: customers_counter,
                latitude: latitude,
                longitude: longitude,
                name: name,
                distances_to_other_customers_map: None,
                time_window_start: 0,
                time_window_end: 0,
                service_time: 0,
                demand: 0,
            };
            customers_vec.push(readed_customer);
            customers_counter += 1;
        }

        return customers_vec;
    }

    fn read_distance_matrix(lines_iter: &mut Lines<BufReader<File>>) -> Option<Vec<Vec<f64>>>{

        let mut distance_matrix: Vec<Vec<f64>> = Vec::new();
        loop {
            let mut readed_line = lines_iter.next().expect("Failed to read next line 1").expect("Failed to read next line 2");
            
            if readed_line.contains("EOF") {
                break;
            }

            let mut distances_parts: Vec<String> = readed_line.split(" ").into_iter().map(|x| x.to_string()).collect();
            distances_parts.pop();
            let distances_vec: Vec<f64> = distances_parts.iter().map(|x| x.parse().unwrap()).collect();
            distance_matrix.push(distances_vec);
        }

        if distance_matrix.len() == 0 {
            return None;
        }

        return Some(distance_matrix);
    }

    pub fn read_customers_demand_info(lines_iter: &mut Lines<BufReader<File>>) -> Vec<Vec<u64>> {

        let mut demand_info: Vec<Vec<u64>> = Vec::new();
        loop {
            let mut readed_line = lines_iter.next().expect("Failed to read next line 1").expect("Failed to read next line 2");
            
            if readed_line.contains("EOF") || readed_line.contains("DEPOT_SECTION"){
                return demand_info;
            }

            readed_line = readed_line.trim().to_string();
            let readed_line: Vec<String> = readed_line.split(" ").into_iter().map(|x| x.to_string()).collect();
            let current_demand_info: Vec<u64> = readed_line.iter().map(|x| x.parse().unwrap()).collect();

            demand_info.push(current_demand_info);
        }
    }

    pub fn read_depot_info(lines_iter: &mut Lines<BufReader<File>>) -> Vec<u64> {

        let mut depot_info: Vec<u64> = Vec::new();
        loop {
            let mut readed_line = lines_iter.next().expect("Failed to read next line 1").expect("Failed to read next line 2");
            
            if readed_line.contains("EOF") || readed_line.contains("-1"){
                return depot_info;
            }

            readed_line = readed_line.trim().to_string();
            let readed_depot_id = readed_line.parse().unwrap();
            depot_info.push(readed_depot_id);
        }
    }

    fn build_distance_matrix(customers_vec: &Vec<Customer>) -> Vec<Vec<f64>> {

        let mut distance_matrix: Vec<Vec<f64>> = Vec::new();
        for i in 0..customers_vec.len() {
            let mut distance_row: Vec<f64> = Vec::new();
            for j in 0..customers_vec.len() {
                let current_distance = customers_vec[i].get_distance_to_other_customer(&customers_vec[j]);
                distance_row.push(current_distance);
            }
            distance_matrix.push(distance_row);
        }

        return distance_matrix;

    }

}

File: examples/vrp_service/python_client/persistence/DomainBuilder.py
import re
import numpy as np

from domain.VehicleRoutingPlan import VehicleRoutingPlan
from domain.Vehicle import Vehicle
from domain.Customer import Customer

class DomainBuilder():

    def __init__(self):
        pass
    
    def build_domain_from_json(self, domain_json):

        dataset_name = domain_json["name"]
        time_windowed = domain_json["time_windowed"]
        distance_matrix = domain_json["distance_matrix"]

        customers_json_values = domain_json["customers_vec"]
        depot_json_values = domain_json["depot_vec"]
        vehicles_json_values = domain_json["vehicles"]

        customers_vec = []
        for i, customer_json in enumerate(customers_json_values):
            customer_i = Customer(
                id = customer_json["id"],
                name = customer_json["name"],
                latitude = customer_json["latitude"],
                longitude = customer_json["longitude"],
                demand = customer_json["demand"],
                time_window_start = customer_json["time_window_start"],
                time_window_end = customer_json["time_window_end"],
                service_time = customer_json["service_time"],
                distances_to_other_customers_dict = None,
            )
            customers_vec.append( customer_i )
        
        depot_vec = []
        for i, depot_json in enumerate(depot_json_values):
            depot_i = Customer(
                id = depot_json["id"],
                name = depot_json["name"],
                latitude = depot_json["latitude"],
                longitude = depot_json["longitude"],
                demand = depot_json["demand"],
                time_window_start = depot_json["time_window_start"],
                time_window_end = depot_json["time_window_end"],
                service_time = depot_json["service_time"],
                distances_to_other_customers_dict = None,
            )
            depot_vec.append( depot_i )
        
        vehicles_vec = []
        for k, vehicle_json in enumerate(vehicles_json_values):
            vehicle_k = Vehicle(
                depot = customers_vec[vehicle_json["depot"]["vec_id"]],
                depot_matrix_id = vehicle_json["depot"]["vec_id"],
                work_day_start = vehicle_json["work_day_start"],
                work_day_end = vehicle_json["work_day_end"],
                capacity = vehicle_json["capacity"],
                customer_list = [customers_vec[customer_json["vec_id"]] for customer_json in vehicle_json["customers"]],
                max_stops = vehicle_json["max_stops"],
            )
            vehicles_vec.append(vehicle_k)

        domain_model = VehicleRoutingPlan(name=dataset_name,
                                          vehicles=vehicles_vec,
                                          customers_dict=customers_vec,
                                          depot_dict=depot_vec,
                                          distance_matrix=distance_matrix,
                                          customers_id_to_matrix_id_map=None,
                                          time_windowed=time_windowed)

        return domain_model


    def build_domain(self, file_path):

        read_file = self._read_vrp_file( file_path )

        dataset_name = read_file["metadata"]["dataset_name"]
        time_windowed = read_file["metadata"]["time_window_task_type"]
        customers_dict = read_file["customers_dict"]
        customers_id_to_matrix_id_map = {}
        for i in range(len(customers_dict)):
            customers_id_to_matrix_id_map[i] = customers_dict[i].id

        if "distance_matrix" in read_file:
            distance_matrix = read_file["distance_matrix"]
            for i in range(len(customers_dict)):
                distances_to_other_customers_dict = {}
                for j in range(len(customers_dict)):
                    current_distance = distance_matrix[i][j]
                    to_customer_name = customers_dict[j].name
                    distances_to_other_customers_dict[to_customer_name] = current_distance
                customers_dict[i].distances_to_other_customers_dict = distances_to_other_customers_dict
        else:
            distance_matrix = self._build_distance_matrix( customers_dict )


        vehicles_capacity = read_file["metadata"]["vehicles_capacity"]
        k_vehicles = read_file["metadata"]["vehicles_count"]
        depot_dict = read_file["depot_dict"]
        n_depots = len(depot_dict)
        max_stops = len(customers_dict) - n_depots

        vehicles_list = []
        for i in range(k_vehicles):
            depot_matrix_id = i % n_depots
            depot = customers_dict[depot_matrix_id]
            work_day_start = depot.time_window_start
            work_day_end = depot.time_window_end
            vehicle = Vehicle(depot, depot_matrix_id, work_day_start, work_day_end,
                              vehicles_capacity, None, max_stops)
            vehicles_list.append( vehicle )

        domain_model = VehicleRoutingPlan(name=dataset_name,
                                          vehicles=vehicles_list,
                                          customers_dict=customers_dict,
                                          depot_dict=depot_dict,
                                          distance_matrix=distance_matrix,
                                          customers_id_to_matrix_id_map=customers_id_to_matrix_id_map,
                                          time_windowed=time_windowed)

        return domain_model

    def _build_distance_matrix(self, customers_dict):


        n_customers = len(customers_dict)
        distance_matrix = np.zeros( (n_customers, n_customers), dtype=np.int64 )
        #distance_matrix = np.zeros((n_customers, n_customers), dtype=np.float32)
        for i in range(n_customers):
            for j in range(n_customers):
                customer_from = customers_dict[i]
                customer_to = customers_dict[j]
                distance_from_to = customer_from.get_distance_to_other_customer( customer_to )
                distance_matrix[i][j] = round(1000 * distance_from_to, 0)
                #distance_matrix[i][j] = distance_from_to

        return distance_matrix

    def _read_vrp_file(self, file_path, for_json=False):

        def read_metadata( file_pointer ):
            metadata = {}
            readed_line = ""

            while True:
                readed_line = file_pointer.readline()
                if "NODE_COORD_SECTION" in readed_line:
                    break

                readed_line = readed_line.replace("\n", "").strip()

                if "NAME" in readed_line:
                    task_name = readed_line.split(" ")[-1]
                    vehicle_count = task_name.split("-")[-1]
                    vehicle_count = int(vehicle_count.replace("k", ""))
                    metadata["vehicles_count"] = vehicle_count

                if "TYPE" in readed_line:
                    task_type = readed_line.split(" ")[-1]
                    task_type = task_type
                    metadata["task_type"] = task_type

                if "NAME" in readed_line:
                    dataset_name = readed_line.split(" ")[-1]
                    metadata["dataset_name"] = dataset_name

                if "EDGE_WEIGHT_TYPE" in readed_line:
                    distance_type = readed_line.split(" ")[-1]
                    distance_type = distance_type
                    metadata["distance_type"] = distance_type

                if "CAPACITY" in readed_line:
                    vehicles_capacity = readed_line.split(" ")[-1]
                    metadata["vehicles_capacity"] = int(vehicles_capacity)

                pass

            return metadata

        def read_customers_common_info( file_pointer, for_json=False ):
            customers_dict = {}
            customer_matrix_id = 0
            readed_line = ""

            while True:
                readed_line = file_pointer.readline()
                if "EOF" in readed_line or "DEMAND_SECTION" in readed_line:
                    break

                readed_line = readed_line.strip()
                readed_line = re.sub(" +", " ", readed_line)
                readed_line = readed_line.split(" ")

                if for_json:
                    id = int( readed_line[0] )
                    latitude = float( readed_line[1] )
                    longitude = float( readed_line[2] )
                    if len(readed_line) > 3:
                        name = readed_line[3].replace("\n", "")
                    else:
                        name = str(id)
                    customers_dict[str(customer_matrix_id)] = {
                        "id": id,
                        "latitude": latitude,
                        "longitude": longitude,
                        "name": name
                    }
                else:
                    id = int( readed_line[0] )
                    latitude = float( readed_line[1] )
                    longitude = float( readed_line[2] )
                    if len(readed_line) > 3:
                        name = readed_line[3].replace("\n", "")
                    else:
                        name = str(id)
                    current_customer = Customer(id=id, latitude=latitude, longitude=longitude, name=name)
                    customers_dict[customer_matrix_id] = current_customer

                customer_matrix_id += 1
            
            customers_dict["n_customers"] = customer_matrix_id

            return customers_dict

        def read_distance_matrix( file_pointer ):
            distance_matrix = []
            readed_line = ""

            while True:
                readed_line = file_pointer.readline()
                if "EOF" in readed_line:
                    break

                readed_line = readed_line.split(" ")
                readed_line.pop(-1)
                matrix_row = [float(value) for value in readed_line]
                matrix_row = np.array( matrix_row, dtype=np.float32 )
                distance_matrix.append( matrix_row )

            distance_matrix = np.array( distance_matrix, dtype=np.float32 )

            return distance_matrix

        def read_customers_demand( file_pointer, customers_dict, for_json=False ):
            customer_matrix_id = 0
            readed_line = ""
            time_window_task_type = False

            while True:
                readed_line = file_pointer.readline()
                if "EOF" in readed_line or "DEPOT_SECTION" in readed_line:
                    break

                readed_line = readed_line.strip()
                readed_line = re.sub(" +", " ", readed_line)
                readed_line = readed_line.split(" ")

                if for_json:
                    customers_dict[str(customer_matrix_id)]["demand"] = float(readed_line[1])
                    if len(readed_line) == 5:
                        time_window_task_type = True
                        customers_dict[str(customer_matrix_id)]["time_window_start"] = int(readed_line[2])
                        customers_dict[str(customer_matrix_id)]["time_window_end"] = int(readed_line[3])
                        customers_dict[str(customer_matrix_id)]["service_time"] = int(readed_line[4].replace("\n", ""))
                else:
                    customers_dict[customer_matrix_id].demand = float(readed_line[1])
                    if len(readed_line) == 5:
                        time_window_task_type = True
                        customers_dict[customer_matrix_id].time_window_start = int(readed_line[2])
                        customers_dict[customer_matrix_id].time_window_end = int(readed_line[3])
                        customers_dict[customer_matrix_id].service_time = int(readed_line[4].replace("\n", ""))

                customer_matrix_id += 1

            return customers_dict, time_window_task_type

        def read_depot_dict( file_pointer, for_json=False ):
            depot_dict = {}
            matrix_id = 0
            readed_line = ""

            while True:
                readed_line = file_pointer.readline()
                if "EOF" in readed_line or "-1" in readed_line:
                    break

                readed_line = readed_line.strip().replace("\n", "")
                depot_id = int(readed_line)
                if for_json:
                    depot_dict[str(matrix_id)] = depot_id
                else:
                    depot_dict[matrix_id] = depot_id
                matrix_id += 1
            
            depot_dict["n_depots"] = matrix_id

            return depot_dict

        read_file_dict = {}
        with open( file_path, "r" ) as vrp_file:
            metadata = read_metadata( vrp_file )
            customers_dict = read_customers_common_info( vrp_file, for_json )

            if metadata["distance_type"] != "EUC_2D":
                distance_matrix = read_distance_matrix( vrp_file )
                read_file_dict["distance_matrix"] = distance_matrix

            customers_dict, time_window_task_type = read_customers_demand( vrp_file, customers_dict, for_json )
            metadata["time_window_task_type"] = time_window_task_type
            depot_dict = read_depot_dict( vrp_file, for_json )

            read_file_dict["metadata"] = metadata
            read_file_dict["customers_dict"] = customers_dict
            read_file_dict["depot_dict"] = depot_dict

        return read_file_dict
